pkg/agent/controller/networkpolicy/reconciler.go

```diff
@@ -386,31 +386,39 @@ func (r *reconciler) registerOFPriorities(rules []*CompletedRule) error {
 	return nil
 }
 
-// add converts CompletedRule to PolicyRule(s) and invokes installOFRule to install them.
+// add converts CompletedRule to PolicyRule(s) and installs them atomically.
+// It persists the realized state to r.lastRealizeds upon success.
 func (r *reconciler) add(rule *CompletedRule, ofPriority *uint16, table binding.TableIDType) error {
 	klog.V(2).Infof("Adding new rule %v", rule)
 	ofRuleByServicesMap, lastRealized := r.computeOFRulesForAdd(rule, ofPriority, table)
+	// Install the flows in a bundle for atomicity.
+	var ofRules []*types.PolicyRule
 	for svcKey, ofRule := range ofRuleByServicesMap {
 		// Each pod group gets an Openflow ID.
 		err := r.idAllocator.allocateForRule(ofRule)
 		if err != nil {
 			return fmt.Errorf("error allocating Openflow ID")
 		}
-		if err = r.installOFRule(ofRule); err != nil {
-			return err
-		}
-		// Record ofID only if its Openflow is installed successfully.
 		lastRealized.ofIDs[svcKey] = ofRule.FlowID
+		ofRules = append(ofRules, ofRule)
 	}
+	if err := r.ofClient.BatchInstallPolicyRuleFlows(ofRules); err != nil {
+		// Do not use idAllocator.forgetRule which is an asynchronous method in case that the actual deletion executes
+		// after the caller retries to add the rule and it succeeds.
+		// As this rule hasn't been installed successfully, it can be removed immediately.
+		for _, rule := range ofRules {
+			r.idAllocator.forgetRuleImmediately(rule.FlowID)
+		}
+		return fmt.Errorf("error installing flows for new rule %v: %v", rule, err)
```

Comment:
I think I am missing something. If the caller retries by calling `add` again, then I would expect `allocateForRule` to be called again for every rule, and a new `FlowID` to be allocated. It shouldn't matter that the deletion of the earlier `FlowID` is asynchronous, since new IDs will be used until that one is released.

Reply:
Thanks for catching this. I missunderstood the key used in idAllocato, thought it was ruleID. Then this change is unnecessary, at least in this bug fix PR.